File: core/tests/unit/project_smoke.cpp
#include <cassert>
#include <filesystem>
#include <iostream>
#include <stdexcept>

#include "streamfind/api.hpp"
#include "streamfind/project.hpp"
#include "../tmp_projects.hpp"

int run() {
    const auto path = streamfind::test::tmp_projects_dir() / "streamfind-core-project-smoke.duckdb";
    std::error_code error;
    std::filesystem::remove(path, error);

    auto project = streamfind::Project::create({path, "test", {{"owner", "test"}}});
    int table_runs = 0;
    int tail_runs = 0;
    streamfind::MethodRegistry registry;
    streamfind::MethodDefinition table_method;
    table_method.id = "test.table";
    table_method.name = "Table";
    table_method.domain = "test";
    table_method.cacheable = true;
    table_method.writes = {"TEST_OUTPUT"};
    registry.register_method(streamfind::Method(
        table_method,
        [&table_runs](streamfind::Project &project, const streamfind::Json &) {
            ++table_runs;
            project.execute_sql("CREATE TABLE IF NOT EXISTS TEST_OUTPUT (value VARCHAR)");
            project.execute_sql("DELETE FROM TEST_OUTPUT");
            project.execute_sql("INSERT INTO TEST_OUTPUT VALUES ('materialized')");
            return streamfind::Json{{"status", "finished"}};
        }));
    streamfind::MethodDefinition tail_method;
    tail_method.id = "test.tail";
    tail_method.name = "Tail";
    tail_method.domain = "test";
    tail_method.cacheable = false;
    registry.register_method(streamfind::Method(
        tail_method,
        [&tail_runs](streamfind::Project &, const streamfind::Json &) {
            ++tail_runs;
            return streamfind::Json{{"status", "tail"}};
        }));
    project.set_metadata({{"owner", "test"}});
    if (project.get_metadata().at("owner") != "test") {
        std::cerr << "metadata getter failed\n";
        return 1;
    }
    project.set_cache("test", "test cache", "hash", {{"value", 42}});
    if (project.get_cache().size() != 1) {
        std::cerr << "cache creation failed\n";
        return 1;
    }
    project.delete_cache();
    if (!project.get_cache().empty()) {
        std::cerr << "cache clear failed\n";
        return 1;
    }
    const auto copy_path = streamfind::test::tmp_projects_dir() / "streamfind-core-project-copy.duckdb";
    std::filesystem::remove(copy_path, error);
    auto copied = project.copy({copy_path, "test", {}});
    if (copied.get_metadata().at("owner") != "test") {
        std::cerr << "project copy failed\n";
        return 1;
    }
    copied.close();
    std::filesystem::remove(copy_path, error);
    streamfind::Workflow table_workflow;
    table_workflow.domain = "test";
    table_workflow.steps.push_back({"test.table", streamfind::ParameterValues{streamfind::Json::object()}});
    project.set_workflow(table_workflow, registry);
    project.run_workflow(registry);
    const auto launch_snapshot = streamfind::Json::parse(project.query_json("SELECT launch_snapshot FROM WORKFLOW_EXECUTION").at(0).at("launch_snapshot").get<std::string>());
    if (launch_snapshot.at("domain") != "test" || launch_snapshot.at("steps").size() != 1 || launch_snapshot.at("steps").at(0).at("id") != "test.table") {
        std::cerr << "workflow launch snapshot was not persisted\n";
        return 1;
    }
    project.execute_sql("UPDATE WORKFLOW_EXECUTION SET status = 'running'");
    project.close();
    project = streamfind::Project::open({path, "test", {}});
    if (streamfind::WorkflowExecutionManager(project).current().at("status") != "interrupted") {
        std::cerr << "workflow restart did not reconcile the active parent\n";
        return 1;
    }
    project.run_workflow(registry);
    if (table_runs != 1) {
        std::cerr << "workflow restart reran a valid completed cached step\n";
        return 1;
    }
    project.execute_sql("INSERT INTO WORKFLOW_EXECUTION_STEP (workflow_revision, step_index, method, parameters, parameter_hash, cache_key, status) VALUES (1, 99, 'stale', '{}', 'stale', 'stale', 'completed')");
    project.execute_sql("DELETE FROM TEST_OUTPUT");
    project.run_workflow(registry);
    if (table_runs != 1 || project.query_json("SELECT COUNT(*) AS count FROM WORKFLOW_EXECUTION_STEP").at(0).at("count") != "1" || project.query_json("SELECT value FROM TEST_OUTPUT") != streamfind::Json::array({{{"value", "materialized"}}})) {
        std::cerr << "cache table materialization failed\n";
        return 1;
    }
    if (project.run_method("test.tail", streamfind::Json::object(), registry).at("status") != "tail" ||
        table_runs != 1 || tail_runs != 1) {
        std::cerr << "workflow tail execution did not reuse the completed prefix\n";
        return 1;
    }
    project.execute_sql("DELETE FROM WORKFLOW_EXECUTION");
    project.delete_cache();
    if (project.run_method("test.tail", streamfind::Json::object(), registry).at("status") != "tail" ||
        table_runs != 2 || tail_runs != 3) {
        std::cerr << "workflow tail execution did not fall back to the full workflow\n";
        return 1;
    }
    const auto execution = project.get_workflow_execution();
    if (execution.at(0).at("status") != "completed") {
        std::cerr << "workflow execution tracking failed\n";
        return 1;
    }
    project.delete_cache();
    const auto execution_table = streamfind::api::run(
        streamfind::api::ProjectCommand::get_workflow_execution,
        {{"database_path", path.string()}});
    if (!execution_table.contains("columns") || execution_table.at("columns").at("step_index").empty()) {
        std::cerr << "workflow execution table API failed\n";
        return 1;
    }
    const auto metadata = streamfind::api::run(
        streamfind::api::ProjectCommand::get_metadata,
        {{"database_path", path.string()}});
    if (streamfind::Json::parse(metadata.at("columns").at("metadata").at(0).get<std::string>()).at("owner") != "test") {
        std::cerr << "metadata API failed\n";
        return 1;
    }
    if (streamfind::api::run(streamfind::api::ProjectCommand::validate,
                             {{"database_path", path.string()}}).at("valid") != true ||
        streamfind::api::run(streamfind::api::ProjectCommand::get_cache_size,
                             {{"database_path", path.string()}}) != 0) {
        std::cerr << "project validation API failed\n";
        return 1;
    }
    project.close();
    auto reopened = streamfind::Project::open({path, "test", {}});
    if (reopened.get_domain() != "test" ||
        reopened.get_metadata().at("owner") != "test") {
        std::cerr << "project reopen failed\n";
        return 1;
    }
    reopened.validate();
    reopened.close();
    std::filesystem::remove(path, error);
    return 0;
}

int main() {
    try {
        return run();
    } catch (const std::exception &error) {
        std::cerr << error.what() << '\n';
        return 1;
    }
}

File: rust/crates/core/tests/project_smoke.rs
use serde_json::{json, Value};
use std::fs;
use std::sync::{
    atomic::{AtomicUsize, Ordering},
    Arc,
};
use std::time::{SystemTime, UNIX_EPOCH};
use streamfind_rust_core::{
    api, Method, MethodRegistry, ParameterSchema, Project, ProjectOptions, Workflow, WorkflowStep,
};

fn install_mass_spec_schema(project: &mut Project) {
    for table in [
        "MASS_SPEC_ANALYSES",
        "MASS_SPEC_SPECTRA_HEADERS",
        "MASS_SPEC_CHROMATOGRAMS_HEADERS",
        "MASS_SPEC_CHROMATOGRAMS",
        "MASS_SPEC_NTA_FEATURES",
        "MASS_SPEC_NTA_SUSPECTS",
        "MASS_SPEC_NTA_INTERNAL_STANDARDS",
        "MASS_SPEC_NTA_TRANSFORMATION_PRODUCTS",
    ] {
        project
            .execute_sql(&format!("CREATE TABLE {table} (marker VARCHAR)"))
            .unwrap();
    }
}

fn temporary_database() -> std::path::PathBuf {
    let stamp = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_nanos();
    streamfind_rust_test_support::tmp_projects_dir().join(format!("streamfind-rust-{stamp}.duckdb"))
}

#[test]
fn project_round_trips_metadata_and_schema() {
    let path = temporary_database();
    let options = ProjectOptions {
        database_path: path.clone(),
        domain: "mass_spec".into(),
        create_if_missing: false,
        read_only: false,
    };
    let mut project = Project::create(options.clone()).unwrap();
    project.set_metadata(json!({"owner": "rust-test"})).unwrap();
    assert_eq!(project.get_metadata()["owner"], "rust-test");
    project
        .set_cache("test", "test cache", "hash", &json!({"value": 42}))
        .unwrap();
    assert_eq!(project.get_cache().unwrap().len(), 1);
    assert!(!project.get_audit_trail().unwrap().is_empty());
    project.delete_cache().unwrap();
    assert!(project.get_cache().unwrap().is_empty());
    install_mass_spec_schema(&mut project);
    let copy_path = temporary_database();
    let copied = project
        .copy(ProjectOptions {
            database_path: copy_path.clone(),
            domain: String::new(),
            create_if_missing: false,
            read_only: false,
        })
        .unwrap();
    assert_eq!(copied.get_metadata()["owner"], "rust-test");
    copied.close();
    fs::remove_file(copy_path).unwrap();
    drop(project);

    let reopened = Project::open(ProjectOptions {
        read_only: true,
        ..options
    })
    .unwrap();
    assert_eq!(reopened.info().domain, "mass_spec");
    assert_eq!(reopened.info().metadata["owner"], "rust-test");
    drop(reopened);
    assert_eq!(
        serde_json::from_str::<serde_json::Value>(
            api::get_metadata(&json!({
                "database_path": path.to_string_lossy()
            }))
            .unwrap()["columns"]["metadata"][0]
                .as_str()
                .unwrap(),
        )
        .unwrap()["owner"],
        "rust-test"
    );
    assert!(api::validate(&json!({
        "database_path": path.to_string_lossy()
    }))
    .unwrap()["valid"]
        .as_bool()
        .unwrap());
    assert_eq!(
        api::get_cache_size(&json!({
            "database_path": path.to_string_lossy()
        }))
        .unwrap(),
        0
    );
    fs::remove_file(path).unwrap();
}

#[test]
fn workflow_cache_restores_written_tables() {
    let path = temporary_database();
    let mut project = Project::create(ProjectOptions {
        database_path: path.clone(),
        domain: "test".into(),
        create_if_missing: false,
        read_only: false,
    })
    .unwrap();
    let runs = Arc::new(AtomicUsize::new(0));
    let runs_for_method = Arc::clone(&runs);
    let tail_runs = Arc::new(AtomicUsize::new(0));
    let tail_runs_for_method = Arc::clone(&tail_runs);
    let mut method = Method::new(
        "test.write_table",
        "test.write_table",
        "Write a test table",
        "test",
        ParameterSchema::default(),
        Box::new(move |project, _| {
            runs_for_method.fetch_add(1, Ordering::SeqCst);
            project.execute_sql(
                "CREATE TABLE IF NOT EXISTS TEST_OUTPUT (row_key VARCHAR, value VARCHAR)",
            )?;
            project.execute_sql("DELETE FROM TEST_OUTPUT WHERE row_key = 'cache-test'")?;
            project.execute_sql("INSERT INTO TEST_OUTPUT VALUES ('cache-test', 'materialized')")?;
            Ok(json!({"status": "finished"}))
        }),
    );
    method.cacheable = true;
    method.writes = vec!["TEST_OUTPUT".into()];
    let mut registry = MethodRegistry::default();
    registry.register(method).unwrap();
    registry
        .register(Method::new(
            "test.tail",
            "test.tail",
            "Append a tail method",
            "test",
            ParameterSchema::default(),
            Box::new(move |_, _| {
                tail_runs_for_method.fetch_add(1, Ordering::SeqCst);
                Ok(json!({"status": "tail"}))
            }),
        ))
        .unwrap();
    project
        .set_workflow(
            Workflow {
                domain: "test".into(),
                steps: vec![WorkflowStep {
                    method: "test.write_table".into(),
                    parameters: json!({}),
                    metadata: None,
                }],
                ..Workflow::default()
            },
            &registry,
        )
        .unwrap();
    let workflow = project.get_workflow().unwrap();
    project
        .run_workflow(&workflow, &registry, None, None)
        .unwrap();
    let launch_snapshot: Value = serde_json::from_str(
        project
            .query_json("SELECT launch_snapshot FROM WORKFLOW_EXECUTION")
            .unwrap()[0]["launch_snapshot"]
            .as_str()
            .unwrap(),
    )
    .unwrap();
    assert_eq!(launch_snapshot["domain"], "test");
    assert_eq!(launch_snapshot["steps"].as_array().unwrap().len(), 1);
    assert_eq!(launch_snapshot["steps"][0]["id"], "test.write_table");
    project
        .execute_sql("UPDATE WORKFLOW_EXECUTION SET status = 'running'")
        .unwrap();
    drop(project);
    let mut project = Project::open(ProjectOptions {
        database_path: path.clone(),
        domain: "test".into(),
        create_if_missing: false,
        read_only: false,
    })
    .unwrap();
    assert_eq!(
        streamfind_rust_core::WorkflowExecutionManager::new(&project)
            .current()
            .unwrap()["status"],
        "interrupted"
    );
    let workflow = project.get_workflow().unwrap();
    project
        .run_workflow(&workflow, &registry, None, None)
        .unwrap();
    assert_eq!(runs.load(Ordering::SeqCst), 1);
    project
        .execute_sql("INSERT INTO WORKFLOW_EXECUTION_STEP (workflow_revision, step_index, method, parameters, parameter_hash, cache_key, status) VALUES (1, 99, 'stale', '{}', 'stale', 'stale', 'completed')")
        .unwrap();
    project
        .execute_sql("DELETE FROM TEST_OUTPUT WHERE row_key = 'cache-test'")
        .unwrap();
    let workflow = project.get_workflow().unwrap();
    project
        .run_workflow(&workflow, &registry, None, None)
        .unwrap();
    assert_eq!(runs.load(Ordering::SeqCst), 1);
    assert_eq!(
        project
            .query_json("SELECT COUNT(*) AS count FROM WORKFLOW_EXECUTION_STEP")
            .unwrap()[0]["count"],
        1
    );
    assert_eq!(
        project
            .query_json("SELECT value FROM TEST_OUTPUT WHERE row_key = 'cache-test'")
            .unwrap(),
        json!([{"value": "materialized"}])
    );
    assert_eq!(
        project
            .run_method("test.tail", &json!({}), &registry)
            .unwrap()["status"],
        "tail"
    );
    assert_eq!(runs.load(Ordering::SeqCst), 1);
    assert_eq!(tail_runs.load(Ordering::SeqCst), 1);
    project
        .execute_sql("DELETE FROM WORKFLOW_EXECUTION")
        .unwrap();
    project.delete_cache().unwrap();
    assert_eq!(
        project
            .run_method("test.tail", &json!({}), &registry)
            .unwrap()["status"],
        "tail"
    );
    assert_eq!(runs.load(Ordering::SeqCst), 2);
    assert_eq!(tail_runs.load(Ordering::SeqCst), 3);
    assert_eq!(
        project.get_workflow_execution().unwrap()[0]["status"],
        "completed"
    );
    let execution_table = api::get_workflow_execution(&json!({
        "database_path": path.to_string_lossy()
    }))
    .unwrap();
    assert_eq!(execution_table["row_count"], 3);
    assert!(execution_table["columns"]["step_index"].is_array());
    fs::remove_file(path).unwrap();
}
